File: 785-is-graph-bipartite/785-is-graph-bipartite.java
class Solution {
    private static boolean isBipartiteDFS(int current, int[][] graph, int[] color) {
        if(color[current] == -1) color[current] = 1;
        for(int adjNode : graph[current]) {
            if(color[adjNode] == -1) {
                color[adjNode] = 1 - color[current];
                if(!isBipartiteDFS(adjNode, graph, color)) 
                    return false;
            } else if(color[adjNode] == color[current]) return false;
        }
        return true;
    }
    public boolean isBipartite(int[][] graph) {
        int n = graph.length;
        int[] color = new int[n];
        for(int i = 0; i < n; ++i) color[i] = -1;
        for(int i = 0; i < n; ++i) 
            if(color[i] == -1 && !isBipartiteDFS(i, graph, color))
               return false;
        return true;
    }
}

File: 785-is-graph-bipartite/785-is-graph-bipartite.py
class Solution:
    def isBipartiteDFS(self, current: int, color: list, graph: List[List[int]]) -> bool:
        if color[current] == -1: 
            color[current] = 1
        for adjNode in graph[current]:
            if color[adjNode] == -1:
                color[adjNode] = 1 - color[current]
                if self.isBipartiteDFS(adjNode, color, graph) == False:
                    return False
            elif color[adjNode] == color[current]:
                return False
        return True
    def isBipartite(self, graph: List[List[int]]) -> bool:
        n = len(graph)
        color = [-1] * n
        for i in range(n):
            if color[i] == -1 and self.isBipartiteDFS(i, color, graph) == False:
                return False
        return True